README.md
#timespanjs
##简介
实现日期差值的人性化显示的javascript库。比如，可以显示成：2年2个月2小时4分钟23秒。同时支持国际化,可扩展。已经支持的语言：
+ 中文简体（zh-cn）
+ 中文繁体（zh-tw）
+ 英文（en）


###如何使用
1. 通过script标签引入timespanjs脚本文件。如果使用requirejs，则需要通过require.config配置timespan路径。
2. 创建Timespan对象。可以通过构造函数中传递时间差值创建，也可以通过传递两个Date对象实例创建。
3. 调用humanize实例方法，获取人性化显示字符串。
4. 调用Timespan.lang()，设置语言。

####浏览器下的引用
```js
<script src="timespan.js"></script>
<script>
    var ts=new Timespan(85,'m');
</script>
```
####requirejs下的引用
```js
require.config({
    paths: {
        "timespanjs": "path/to/timespan",
    }
});
define(["timespanjs"], function (Timespan) {
    var ts=new Timespan(85,'m');
});
```
####node.js下的引用
安装nodejs，通过npm命令安装timespanjs（注意：由于timespanjs被抢注，这里采用timespan-js作为nodejs中的包名）。npm地址：[timespan-js](https://www.npmjs.org/package/timespan-js)

```js
npm install timespan-js
```
调用方法：
```js
var Timespan=require("timespan-js");
var ts=new Timespan(85,"m");
console.log(ts.humanize());
```

####调用示例：
```js
//该值表示为：3小时4分钟25秒30毫秒
var msvalue = 30 + 1000 * 25 + 1000 * 60 * 4 + 1000 * 60 * 60 * 3;
var ts = new Timespan(msvalue, 'ms');
console.log(ts.humanize());//输出为：3小时4分钟25秒30毫秒

//通过fromDates方法构造
var dt=new Date(2014,7,1);
var dt1=new Date(2014,8,1,10,12,15,234);
var ts = Timespan.fromDates(dt, dt1);
console.log(ts.humanize());
//输出：1个月1天10小时12分钟15秒，234毫秒未输出，因为baseUnit参数默认是:'s'
Timespan.lang('en');
console.log(ts.humanize());
//输出：1 month,1 day,10 hours,12 minutes,15 seconds

```
##API说明
见[api说明](https://github.com/houyhea/timespanjs/blob/master/doc/api.md)。
##测试用例
见[https://github.com/houyhea/timespanjs/blob/master/test/testTimespan.html](https://github.com/houyhea/timespanjs/blob/master/test/testTimespan.html)。请获取并运行即可。
##浏览器兼容性
兼容IE8+,chrome,firefox。
##依赖
不需要依赖其他库。
##协议
采用[MIT 许可协议](https://github.com/houyhea/timespanjs/blob/master/LICENSE)。
##帮助
支付宝赞助（houyhea）：  
![赞助](https://raw.githubusercontent.com/houyhea/lab/master/alipayqrcode.png)

History.md
# 开发日志

## v0.1.1
1.  增加get方法；
2.  扩展config方法；
## v0.1.0
1.  扩展add、substract、set方法功能，支持传递timespan对象作为参数。
2.  增加nodejs支持，发布到npm。地址：[timespan-js](https://www.npmjs.org/package/timespan-js)


## v0.0.2
1. 完善主体代码逻辑
2. 添加History.md
3. 添加LICENSE(MIT)
4. 添加api.md
## v0.0.1
1. 创建库
2. 添加基本文件  
3. 编写基础代码




timespan.js
/**
 * Created by houyhea on 14-8-20.
 * copyright © houyhea{at}126.com www.timespanjs.com
 *
 */
(function () {
    'use strict';
    var VERSION = "0.1.1",
        globalScope = typeof global !== 'undefined' ? global : this,
        oldGlobalMoment,
        languages = {},
        MILLISECONDS_PER_SECOND = 1000,
        SECONDS_PER_MINUTE = 60,
        MINUTES_PER_HOUR = 60,
        HOURS_PER_DAY = 24,
        DAYS_PER_WEEK = 7,
        DAYS_PER_MONTH = 30,    //由于这里体现的是相对值（差值），所以这里统一规定：一个月就是30天
        MONTHS_PER_YEAR = 12,    //统一规定，一年360天。
        CONFIG = {
            digits: 2,         //小数点位数
            baseUnit: "s",     //最小显示单位，在现有单位中配置，从 y 到 ms
            length: 0          //友好字符串显示组数，如果为0，则显示全部（如果某个单位上为0，则不计入显示）
        };

    /*******************************************************
     private functions
     *******************************************************/
    function getMilliseconds(value, unit) {
        var ret = value;
        switch (unit) {
            case "milliseconds":
            case "ms":
                ret = value;
            default :
                break;
            case "seconds":
            case "s":
                ret = value * MILLISECONDS_PER_SECOND;
                break;
            case "minutes":
            case "m":
                ret = value * MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE;
                break;
            case "hours":
            case "h":
                ret = value * MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR;
                break;
            case "days":
            case "d":
                ret = value * MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY;
                break;
            case "weeks":
            case "w":
                ret = value * MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY * DAYS_PER_WEEK;
                break;
            case "months":
            case "M":
                ret = value * MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY * DAYS_PER_MONTH;
                break;
            case "years":
            case "y":
                ret = value * MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY * DAYS_PER_MONTH * MONTHS_PER_YEAR;
                break;
        }
        return ret;
    }

    function extend(a, b) {
        if (!b || !a)
            return a;

        for (var i in b) {
            if (b.hasOwnProperty(i)) {
                a[i] = b[i];
            }
        }

        if (b.hasOwnProperty("toString")) {
            a.toString = b.toString;
        }

        if (b.hasOwnProperty("valueOf")) {
            a.valueOf = b.valueOf;
        }

        return a;
    }

    function getLang(key) {
        return languages[key];
    }

    function loadLang(key, value) {
        value.abbr = key;
        languages[key] = value;
        return languages[key];
    }

    function _humanize(ts, baseUnit, length, delimiter) {
        var methods = "years_months_weeks_days_hours_minutes_seconds_milliseconds".split("_");
        var units = "y_M_w_d_h_m_s_ms".split("_");
        var i = units.indexOf(baseUnit);
        if (i < 0)i = units.length - 2;//默认：秒
        if (length > units.length)length = units.length;
        if (length <= 0)length = units.length;
        var texts = [];
        var len = 0;
        for (var j = i; j >= 0; j--) {
            if (len >= length)
                break;
            var v = ts[methods[j]]();
            if (v <= 0)
                continue;

            var t = (v > 1 ? this.PLURAL[units[j]] : this.SINGLUAR[units[j]]).replace(/%d/i, v);
            texts.push(t);
            len++;
        }
        return texts.reverse().join(delimiter || "");
    }

    /******************************************************
     Timespan class
     *******************************************************/

    /**
     *构造函数
     * @param value 必须，number类型。时间差值
     * @param unit 可选，指定value的单位。默认：ms，可选值：y|years,M|months,d|days,h|hours,m|minutes,s|seconds,ms|milliseconds
     * @param config 可选，配置参数。
     * @constructor
     */
    var Timespan = function (value, unit, config) {
        if (!(value instanceof Number)) {
            throw new Error("value must be a number");
        }
        unit = unit || "ms";
        this.msec = getMilliseconds(value, unit);
        this._config = extend(CONFIG, config);
    }

    /**
     * 通过javascript Date对象获取Timespan实例
     * @param startDate 起始日期，必须
     * @param endDate 结束日期，可选。默认:now
     * @returns {Timespan}
     */
    Timespan.fromDates = function (startDate, endDate) {
        endDate = endDate || Date.now();
        var msec = Math.abs(startDate.getTime() - endDate.getTime());
        return new Timespan(msec, 'ms');
    }

    Timespan.version = VERSION;
    /**
     * 设置当前语言（如果value为空），添加语言（如果value不为空）
     * @param key 语言键
     * @param value 语言对象
     * @returns {string|n.abbr|*|abbr}
     */
    Timespan.lang = function (key, value) {
        if (!key) {
            return Timespan.prototype._lang.abbr;
        }
        if (value) {
            var l = loadLang(key, value);
        } else {
            var r = Timespan.prototype._lang = getLang(key);
            return r.abbr;
        }

    };

    Timespan.prototype = {
        milliseconds: function () {
            return this.msec % MILLISECONDS_PER_SECOND;
        },
        seconds: function () {
            return Math.floor(this.msec / MILLISECONDS_PER_SECOND) % SECONDS_PER_MINUTE;
        },
        minutes: function () {
            return Math.floor(this.msec / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE)) % MINUTES_PER_HOUR;
        },
        hours: function () {
            return Math.floor(this.msec / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR)) % HOURS_PER_DAY;
        },
        days: function () {
            var days = Math.floor(this.msec / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY)) % DAYS_PER_MONTH;
            if (days >= DAYS_PER_WEEK)
                return days % DAYS_PER_WEEK;
            else
                return days;
        },
        weeks: function () {
            var days = Math.floor(this.msec / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY)) % DAYS_PER_MONTH;
            return Math.floor(days / DAYS_PER_WEEK);
        },
        months: function () {
            return Math.floor(this.msec / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY * DAYS_PER_MONTH)) % MONTHS_PER_YEAR;
        },
        years: function () {
            return Math.floor(this.msec / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY * DAYS_PER_MONTH * MONTHS_PER_YEAR));
        },
        asMilliseconds: function () {
            return +this;
        },
        asSeconds: function () {
            return (+this / MILLISECONDS_PER_SECOND).toFixed(this._config.digits);
        },
        asMinutes: function () {
            return (+this / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE)).toFixed(this._config.digits);
        },
        asHours: function () {
            return (+this / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR)).toFixed(this._config.digits);
        },
        asDays: function () {
            return (+this / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY)).toFixed(this._config.digits);
        },
        asWeeks: function () {
            return (+this / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY * DAYS_PER_WEEK)).toFixed(this._config.digits);

        },
        asMonths: function () {
            return (+this / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY * DAYS_PER_MONTH)).toFixed(this._config.digits);
        },
        asYears: function () {
            return (+this / (MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY * DAYS_PER_MONTH * MONTHS_PER_YEAR)).toFixed(this._config.digits);
        },
        humanize: function (baseUnit, length) {
            return this.lang().humanize(this, baseUnit || this._config.baseUnit, length || this._config.length);
        },
        add: function (value, unit) {
            if (!value)
                return;
            if (value instanceof Timespan) {
                this.msec += value.asMilliseconds();
                return;
            }
            var ms = getMilliseconds(value, unit);
            this.msec += ms;
        },
        subtract: function (value, unit) {
            if (!value)
                return;
            if (value instanceof Timespan) {
                this.msec -= value.asMilliseconds();
                return;
            }
            var ms = getMilliseconds(value, unit);
            this.msec -= ms;
        },
        get: function () {
            return this.msec;
        },
        set: function (value, unit) {
            if (!value)
                return;
            if (value instanceof Timespan) {
                this.msec = value.asMilliseconds();
                return;
            }
            var ms = getMilliseconds(Math.abs(value), unit);
            this.msec = ms;
        },
        config: function (config) {
            if (config === undefined) {
                return this._config;
            }
            else {
                extend(this._config, config);
            }
        },
        lang: function (key) {
            if (key === undefined) {
                return this._lang;
            } else {
                this._lang = getLang(key);
                return this;
            }
        },
        toString: function () {
            return this.msec + '';
        },
        valueOf: function () {
            return this.msec;
        }

    };


    /******************************************************
     languages
     *******************************************************/

    /**
     * 中文-简体(zh-cn)
     */
    (function (factory) {
        factory(Timespan);
    }(function (Timespan) {
        return Timespan.lang('zh-cn', {
            humanize: function (ts, baseUnit, length) {
                return _humanize.call(this, ts, baseUnit, length, this.DELIMITER);

            },
            DELIMITER: "",
            SINGLUAR: {
                ms: "%d毫秒",
                s: "%d秒",
                m: "%d分钟",
                h: "%d小时",
                d: "%d天",
                w: "%d周",
                M: "%d个月",
                y: "%d年"
            },
            PLURAL: {
                ms: "%d毫秒",
                s: "%d秒",
                m: "%d分钟",
                h: "%d小时",
                d: "%d天",
                w: "%d周",
                M: "%d个月",
                y: "%d年"
            }
        });
    }));
    /**
     * 中文-繁体(zh-tw)
     */
    (function (factory) {
        factory(Timespan);
    }(function (Timespan) {
        return Timespan.lang('zh-tw', {
            humanize: function (ts, baseUnit, length) {
                return _humanize.call(this, ts, baseUnit, length, this.DELIMITER);
            },
            DELIMITER: "",
            SINGLUAR: {
                ms: "%d毫秒",
                s: "%d秒",
                m: "%d分鐘",
                h: "%d小時",
                d: "%d天",
                w: "%週",
                M: "%d個月",
                y: "%d年"
            },
            PLURAL: {
                ms: "%d毫秒",
                s: "%d秒",
                m: "%d分鐘",
                h: "%d小時",
                d: "%d天",
                w: "%d週",
                M: "%d個月",
                y: "%d年"
            }
        });
    }));
    /**
     * 英文(en)
     */
    (function (factory) {
        factory(Timespan);
    }(function (Timespan) {
        return Timespan.lang('en', {
            humanize: function (ts, baseUnit, length) {
                return _humanize.call(this, ts, baseUnit, length, this.DELIMITER);
            },
            DELIMITER: ",",
            SINGLUAR: {
                ms: "%d ms",
                s: "%d second",
                m: "%d minute",
                h: "%d hour",
                d: "%d day",
                w: "%d week",
                M: "%d month",
                y: "%d year"
            },
            PLURAL: {
                ms: "%d ms",
                s: "%d seconds",
                m: "%d minutes",
                h: "%d hours",
                d: "%d days",
                w: "%d weeks",
                M: "%d months",
                y: "%d years"
            }
        });
    }));
    Timespan.lang("zh-cn");//设置默认语言

    Timespan.noConflict = function () {
        globalScope.Timespan = oldGlobalMoment;
        return Timespan;
    }
    if (typeof module !== 'undefined' && module.exports) {
        module.exports = Timespan;

    } else if (typeof window.define === 'function' && window.define.amd) {
        window.define('timespanjs', [], function () {
            return Timespan;
        });
    }
    else {
        oldGlobalMoment = globalScope.Timespan;
        globalScope.Timespan = Timespan;
    }

    return Timespan;

}).call(this);